<a href="https://colab.research.google.com/github/Wangari-Kariuki/Program-interest-vs-graduation-rate/blob/main/final_project_report.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Part 1:
**Exploring data of enrollment students from 2017 to 2021**

This project involves a descriptive analysis of student academic interest information.

The data sets contain the following information.

A list of enrolling applicants (fall, degree-seeking, first-year students only) from the years 2017-2021.  There is a column for their Primary Academic Interest from their application, and a column listing all academic interests they marked on their application.

A list of all programs attached to each ID number. “Program” is the degree level (BA for Bachelor’s degree) attached to a Major code (CS for Computer Science).

The program status, and the start and end date of each program.  

The major(s) at graduation (or current major(s) if the student is still enrolled) and the current enrollment status

 Objective of this projct is to examine whether there is a correlation between  student's major choices and their probability of switching majors. Additionally, we will track the graduation status of students who switch majors.
We will do this by keeping the following qestions in mind during the  analysis
1. What students have switched from one major to another?
2. What majors do most students switch to and from?
3. What is the graduations status of students who switch majors?

Visualizations may inlude:
1. Time trend of students from listing major options, dropping, switching and  finally graduating.
2. Bar graph or histogram of different majors and number of students who express interest of them.




**Importing the data sets**

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

original_total_enrolled = pd.read_excel('/content/drive/MyDrive/CS215-Spring-ESTHER/Projects/Final Project/Enrolled Applicants fall 2017-2021.xlsx')
original_total_enrolled.columns
original_total_enrolled.info()
original_total_enrolled.count()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2010 entries, 0 to 2009
Data columns (total 4 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   Student ID                 2010 non-null   int64 
 1   Start Term                 2010 non-null   object
 2   All Academic Interests     2010 non-null   object
 3   Primary Academic Interest  1939 non-null   object
dtypes: int64(1), object(3)
memory usage: 62.9+ KB


,0
Student ID,2010
Start Term,2010
All Academic Interests,2010
Primary Academic Interest,1939


In [2]:
grad_df = pd.read_csv('/content/drive/MyDrive/CS215-Spring-ESTHER/Projects/Final Project/Enrolled Applicants fall 2017-2021.xlsx - Declared Majors.csv')
grad_df.info()
grad_df.head(4)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4444 entries, 0 to 4443
Data columns (total 9 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   Student ID                 4444 non-null   int64 
 1   Student Status             4444 non-null   object
 2   Start Term                 4444 non-null   object
 3   Program Code               4444 non-null   object
 4   Program Status             4431 non-null   object
 5   Prog Add Date              4444 non-null   object
 6   Prog Status Date           4444 non-null   object
 7   Majors at Graduation       3554 non-null   object
 8   Currently Declared Majors  153 non-null    object
dtypes: int64(1), object(8)
memory usage: 312.6+ KB


,Student ID,Student Status,Start Term,Program Code,Program Status,Prog Add Date,Prog Status Date,Majors at Graduation,Currently Declared Majors
0,2649639,Graduated,2017FA,UNDC,Declared,05/05/2017,01/18/2019,THTR,NaN
1,2649639,Graduated,2017FA,BA.THTR,Graduated,01/18/2019,08/27/2021,THTR,NaN
2,5423674,Withdrawn,2017FA,UNDC,Withdrawn,02/19/2003,06/01/2018,NaN,NaN
3,7077890,Graduated,2020FA,UNDC,Declared,08/24/2020,11/12/2020,SOC,NaN


Filtering out the students who withrew from whitman

In [3]:
#let's se how many students withdrew
withdrawn = grad_df.loc[grad_df['Student Status'] ==  "Withdrawn"]
withdrawn.info()
print(f"{len(withdrawn)} students withdrew  within 2017 and 2021 academic years ")

<class 'pandas.core.frame.DataFrame'>
Index: 737 entries, 2 to 4441
Data columns (total 9 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   Student ID                 737 non-null    int64 
 1   Student Status             737 non-null    object
 2   Start Term                 737 non-null    object
 3   Program Code               737 non-null    object
 4   Program Status             734 non-null    object
 5   Prog Add Date              737 non-null    object
 6   Prog Status Date           737 non-null    object
 7   Majors at Graduation       0 non-null      object
 8   Currently Declared Majors  0 non-null      object
dtypes: int64(1), object(8)
memory usage: 57.6+ KB
737 students withdrew  within 2017 and 2021 academic years 


In [4]:
#dropping withdrawn students from original enrolled students df by matching their student ids
w_ids = withdrawn['Student ID']
#dropping withdrawn students from the first df
current_enrollment_info= original_total_enrolled[~original_total_enrolled['Student ID'].isin(w_ids)]
current_enrollment_info.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1628 entries, 0 to 2009
Data columns (total 4 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   Student ID                 1628 non-null   int64 
 1   Start Term                 1628 non-null   object
 2   All Academic Interests     1628 non-null   object
 3   Primary Academic Interest  1573 non-null   object
dtypes: int64(1), object(3)
memory usage: 63.6+ KB


Now we have obtained initial enrollement  infromation for current students. Next we will obtain the graduation information for those current students.Note that grad infromation means the lst of majors they declared throughout the 4 years of college.

In [5]:
#dropping withdrawn students' information from second df
current_grad_info = grad_df.drop(withdrawn.index)
current_grad_info .info()


<class 'pandas.core.frame.DataFrame'>
Index: 3707 entries, 0 to 4443
Data columns (total 9 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   Student ID                 3707 non-null   int64 
 1   Student Status             3707 non-null   object
 2   Start Term                 3707 non-null   object
 3   Program Code               3707 non-null   object
 4   Program Status             3697 non-null   object
 5   Prog Add Date              3707 non-null   object
 6   Prog Status Date           3707 non-null   object
 7   Majors at Graduation       3554 non-null   object
 8   Currently Declared Majors  153 non-null    object
dtypes: int64(1), object(8)
memory usage: 289.6+ KB


In [6]:
#our data contains two sets of information fro each student, first is when they declared their major and second is when so we separate the two sets of information into two datasets
declared = current_grad_info.drop_duplicates(subset=['Student ID'], keep='last')
declared.info()
declared.head(2)

<class 'pandas.core.frame.DataFrame'>
Index: 1717 entries, 1 to 4443
Data columns (total 9 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   Student ID                 1717 non-null   int64 
 1   Student Status             1717 non-null   object
 2   Start Term                 1717 non-null   object
 3   Program Code               1717 non-null   object
 4   Program Status             1716 non-null   object
 5   Prog Add Date              1717 non-null   object
 6   Prog Status Date           1717 non-null   object
 7   Majors at Graduation       1648 non-null   object
 8   Currently Declared Majors  69 non-null     object
dtypes: int64(1), object(8)
memory usage: 134.1+ KB


,Student ID,Student Status,Start Term,Program Code,Program Status,Prog Add Date,Prog Status Date,Majors at Graduation,Currently Declared Majors
1,2649639,Graduated,2017FA,BA.THTR,Graduated,01/18/2019,08/27/2021,THTR,NaN
4,7077890,Graduated,2020FA,BA.SOC,Graduated,11/12/2020,06/01/2023,SOC,NaN


**Merging with graduation data**

We merge the post enrollement info with the enrollement info  of retained students

In [7]:
enrollement_and_grad = current_enrollment_info.merge(declared, on = "Student ID", how= "left")
enrollement_and_grad.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1628 entries, 0 to 1627
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   Student ID                 1628 non-null   int64 
 1   Start Term_x               1628 non-null   object
 2   All Academic Interests     1628 non-null   object
 3   Primary Academic Interest  1573 non-null   object
 4   Student Status             1628 non-null   object
 5   Start Term_y               1628 non-null   object
 6   Program Code               1628 non-null   object
 7   Program Status             1627 non-null   object
 8   Prog Add Date              1628 non-null   object
 9   Prog Status Date           1628 non-null   object
 10  Majors at Graduation       1562 non-null   object
 11  Currently Declared Majors  66 non-null     object
dtypes: int64(1), object(11)
memory usage: 152.8+ KB


**Evaluating initial major interests**

Although this was a tergent from examining the trend in major change, it later became usedul in evaluating the variation in level of interest for different  majors at enrollment.

In [8]:
noPrimaryAcademicInterest = enrollement_and_grad.loc[enrollement_and_grad['Primary Academic Interest'] == 'Undecided']
noPrimaryAcademicInterest  #primary academic interest not stated
print(f"Number of students with no listed primary academic interest: {len(noPrimaryAcademicInterest)}")

Number of students with no listed primary academic interest: 107


In [9]:
noInterest = enrollement_and_grad.loc[enrollement_and_grad['All Academic Interests']== 'Undecided']
noInterest #students with no stated academic interests
print(f" Number of students with no listed secondary academic interests: {len(noInterest)}")

 Number of students with no listed secondary academic interests: 40


In [10]:
#now let's see how many students have neither primary nor secondary academic interests
no_aca_int = enrollement_and_grad[(enrollement_and_grad['All Academic Interests'] == 'Undecided') & (enrollement_and_grad['Primary Academic Interest'] == 'Undecided')]
print(f"Number of students with no stated academic interests in either categories: {len(no_aca_int)}")

Number of students with no stated academic interests in either categories: 22


In [11]:
# This step is crucial if the cell was run multiple times and the column became garbled
enrollement_and_grad['All Academic Interests'] =enrollement_and_grad.loc[enrollement_and_grad.index, 'All Academic Interests']

# Ensure all entries are strings before splitting
enrollement_and_grad['All Academic Interests'] = enrollement_and_grad['All Academic Interests'].astype(str)
# Replace specific 'Nan' string if it exists after conversion; also consider 'nan' (lowercase) for np.nan
enrollement_and_grad['All Academic Interests'] = enrollement_and_grad['All Academic Interests'].replace('Nan', '')
# Now split the strings and strip whitespace from each item, filtering out any empty strings
enrollement_and_grad['All Academic Interests'] = enrollement_and_grad['All Academic Interests'].str.split(',').apply(lambda x: [item.strip() for item in x if item.strip()])
print(type(enrollement_and_grad['All Academic Interests'].iloc[0])) # we've successfully converted each item in our series into a list

<class 'list'>


In [12]:
enrollement_and_grad['count'] = enrollement_and_grad['All Academic Interests'].str.len()
enrollement_and_grad[['count', 'All Academic Interests']]
print(f"Average number of programs students show secondary  interest in : {enrollement_and_grad['count'].mean()}")

Average number of programs students show secondary  interest in : 4.086609336609337


In [13]:
#we split the students according to those with above average number of interests and those with below average number of s. interests
many_interests = enrollement_and_grad.loc[enrollement_and_grad['count'] >= 4]
many_interests
print(f"{len(many_interests)} students have more interests than average")
#what is the modal number of s. interests
print(f"most students list {len(enrollement_and_grad['count'].mode())} intersts ") #positively skewed variance of number of interests per student

849 students have more interests than average
most students list 1 intersts 


In [14]:
few_interests = enrollement_and_grad.loc[enrollement_and_grad['count'] < 4]
print(f"{len(few_interests)} students have less interests than average")

779 students have less interests than average


In [15]:
import plotly.express as px

# Create the histogram using Plotly Express
fig = px.histogram(enrollement_and_grad, x='count', nbins=20)

# Update layout for aesthetics similar to original matplotlib plot
fig.update_layout(
    title_text="Distribution of students within a range of 1 to 20 academic interests",
    xaxis_title_text="number of interest",
    yaxis_title_text="students",
    xaxis_range=[0, 20],
    yaxis_range=[0, 400],
    bargap=0.01 # Adjust for edgecolor effect, creating small gaps between bars
)

# Save the interactive HTML plot
fig.write_html("first_visuali.html")

# Display the plot
fig.show()

Majority of the students listed 3 subjects of interests during their first year if enrollement in college. This information is useful since we could match this students along with their graduation statuses.

# Part 2

Here, things get even more interesting! We will compare the major interest at enrollement and within the 4 years of college for retained students.

This step involves using mapping pipelines to obtain program codes ie(BIOL for Biology) for listed enrollment interest because, as stated earlier, enrollment interest is in the form of simple program titles.

We need them to be in the same format because the colum of graduation majors and swithced majors is in from of program codes.


To achieve this we'll use a library called fuzzy wuzzy. A limtation of this library for our task is that it does character matching which is inefficient hre since our analysis requires the program to identify the meanings and not just character match the strings. To create a structured and constrained pipeline to ensure maximum accuracy of our program we follow these steps:

1. Normalization
2. Creating a controlled vocabulary map
3. Passing the lists of strings through fuzzy pipeline and mapping dictionary



1. Normalization

In [16]:
#importing and evaluating the format of program codes dataset
program_codes = pd.read_csv('/content/drive/MyDrive/CS215-Spring-ESTHER/Projects/Final Project/Enrolled Applicants fall 2017-2021.xlsx - Program Codes and Titles.csv')
type(program_codes['Program Title'].iloc[0])
program_codes['Program Title'].iloc[0]
type(enrollement_and_grad['All Academic Interests'].iloc[0][0])
type(program_codes['Program Title'].iloc[0])
#int_dec.columns

str

In [17]:
#we need some normalization
import re #this module cleans data by removing unwanted text
def normalize(text):
  text = text.lower()
  text = re.sub(r"[^a-z0-9\s]", "", text) # remove punctuations
  text = re.sub(r"\s+", " ", text).strip() #fix spacing
  return text

program_codes['normalized_titles'] = program_codes['Program Title'].apply(normalize)
enrollement_and_grad['normalized_interests'] = enrollement_and_grad['All Academic Interests'].apply(lambda x: [normalize(interest) for interest in x])

2. Creating a controlled vocabulary map

In [18]:
# Create mapping dictionary
interests = enrollement_and_grad['normalized_interests']
program_map = dict(
    zip(
        program_codes['normalized_titles'],
        program_codes['Program Code']
    )
)

3. Mapping lists of strings to

In [19]:
# Install fuzzywuzzy library for fuzzy string matching
#Allows to match the subjects despite the inconsistent naming order compared to the program code titles

!pip install fuzzywuzzy
!pip install python-Levenshtein

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 153.3/153.3 kB 2.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.1/3.1 MB 19.5 MB/s eta 0:00:00


In [20]:
from fuzzywuzzy import process

def  match_interest_to_code(interests_list, program_map, threshold=50):
    matched_codes = []
    for interest in interests_list:
        # Find the best matching program title for the given interest
        # using fuzzy matching. process.extractOne returns (match, score)
        match = process.extractOne(interest, program_map.keys(), score_cutoff=threshold)
        if match:
            # If a match is found above the threshold, add its corresponding code
            matched_codes.append(program_map[match[0]])
    return list(set(matched_codes)) # Return unique codes

# Apply the fuzzy matching function to create new 'fuzzy interest codes'
enrollement_and_grad['fuzzy_interest_codes'] = enrollement_and_grad['normalized_interests'].apply(
    lambda x: match_interest_to_code(x, program_map)
)

# Display the results with both normalized interests and the new fuzzy codes
enrollement_and_grad[['normalized_interests',  'fuzzy_interest_codes', 'Program Code']]
type(enrollement_and_grad['fuzzy_interest_codes'].iloc[0])

list

In [21]:
#removing the first two characters BA in Program codes
enrollement_and_grad['Program Code'] = enrollement_and_grad['Program Code'].str.strip('BA.')
enrollement_and_grad[['normalized_interests',  'fuzzy_interest_codes', 'Program Code', 'count']].head(5)

,normalized_interests,fuzzy_interest_codes,Program Code,count
0,"[biochemistry biophysics molecular biology, en...","[AR.EN, BBMB, BB.PE]",PH.EN,3
1,"[politics, prelaw, undecided]","[UNDC, LCH, POL]",SOC,3
2,"[anthropology, biology, computer science, econ...","[ECON, PSYC, IRES, PHIL, CS, LCH, ANTH, SOC, B...",PSYC,13
3,"[anthropology, art history, economics, environ...","[ECON, PHIL, ARTH, AR.EN, ANTH]",ECON,5
4,"[anthropology, english, environmental studies,...","[JPNS, AR.EN, HIST, ANTH, ENGL]",NTH,5


In [22]:
type(enrollement_and_grad['Program Code'].iloc[0])

str

# Part 3

Hurray! We're done with most of the dirty work now we'll make some cool visualizations to represent the variation in student count per major in enrollment data that we normalized in part 2.

In [122]:
# What was the rate of interest for the 143 possible program choices?
#to do this we will count the number of students who listed each unique program and represent that ina  barchart.
# iterate through each unique studetID and count how many students hve listed program code[''program codes]
import pandas as pd
import numpy as np
def count_interests(current_program_code):
  count = 0
  #iterating through the interest_codes
  for student_fuzzy_codes in enrollement_and_grad['fuzzy_interest_codes']:
    if current_program_code in student_fuzzy_codes:
      count += 1
  return count
program_codes['enrollement_interest_Count'] = program_codes['Program Code'].apply(count_interests)
print("Program codes with theri interest counts (Top 5):")
print(program_codes[['Program Code', 'Program Title', 'enrollement_interest_Count']].sort_values(by = 'enrollement_interest_Count', ascending = False).head())

Program codes with theri interest counts (Top 5):
    Program Code                    Program Title  enrollement_interest_Count
20          BIOL                          Biology                         433
4          AR.EN        Art-Environmental Studies                         429
100          NMS  Natural & Mathematical Sciences                         428
109         PSYC                       Psychology                         394
128         UNDC                       Undeclared                         363


In [123]:
null_codes_indices = np.where(program_codes['enrollement_interest_Count'] == 0)[0] # retunrs the array insidethe tuple generated by np.where()
print(f"Number of programs with no enrollment interest: {len(null_codes_indices)}")
print(f"Indices of programs with no enrollment interest: {null_codes_indices}")

enrollement_interest_PC = program_codes.drop(null_codes_indices)
enrollement_interest_PC.info()

Number of programs with no enrollment interest: 95
Indices of programs with no enrollment interest: [  2   5   7   9  14  16  17  18  19  21  22  23  24  28  30  32  33  34
  35  37  38  39  40  42  44  45  46  47  48  50  51  52  53  54  55  56
  58  62  63  66  67  68  69  70  71  72  73  74  79  80  81  82  83  85
  86  87  88  89  91  92  93  95  96  97  98  99 101 102 103 106 108 110
 111 115 116 117 119 120 121 122 125 127 129 130 131 132 133 134 135 136
 137 138 139 140 141]
<class 'pandas.core.frame.DataFrame'>
Index: 47 entries, 0 to 128
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Program Code                47 non-null     object 
 1   Program Title               47 non-null     object 
 2   normalized_titles           47 non-null     object 
 3   enrollement_interest_Count  47 non-null     int64  
 4   first_choice_programs       47 non-null     int64  
 5   grad_majors

Plotting a categorical graph of student count per major

In [124]:
import plotly.express as px

# Sort program_codes by 'Count' for better visualization
enrollement_interest_PC_sorted = enrollement_interest_PC.sort_values(by='enrollement_interest_Count', ascending=False)

#total number of students with interests for percentage calculation
total_students_with_interests = enrollement_interest_PC_sorted['enrollement_interest_Count'].sum()

#percentage for each program
enrollement_interest_PC_sorted['Percentage'] = (enrollement_interest_PC_sorted['enrollement_interest_Count'] / total_students_with_interests * 100).round(2)
num_programs = enrollement_interest_PC_sorted['Program Code'].count()
# the horizontal bar chart using Plotly Express
fig = px.bar(
    enrollement_interest_PC_sorted,
    y='Program Code',
    x='enrollement_interest_Count',
    orientation='h', # Ensures horizontal bar chart
    title='Distribution of Interest for Programs',
    labels={'Program Code': 'Program Type', 'Count': 'Number of Students'},
    hover_data=['Program Title', 'enrollement_interest_Count', 'Percentage'] # Include Program Title, Count, and Percentage in hover info
)

# Adjust layout for better readability (optional, Plotly usually handles this well)
fig.update_layout(
    yaxis={'categoryorder': 'total ascending'}, # Order bars by count in ascending order for better visual flow
    xaxis_title="Programs",
    yaxis_title="number of students",
    height=max(600, num_programs * 15), # Dynamically adjust height for better visibility of all bars
    margin=dict(l=100, r=20, t=50, b=50) # Adjust margins if needed
)
fig.write_html("second_visuali.html")
# Display the interactive plot
fig.show()

# Part 4
This is now the heart o our analysis;Examining the subset of tudents who switched majors. We will answer the research questions listed in part 1.

1. Extracting enrollement information about students who swithced majors

In [62]:
#to do tis we will extract IDs of students whos rogram status reads changed mind
changed_mind = current_grad_info[current_grad_info['Program Status'] == "Changed Mind"]
# Identify unique student IDs who ever had a 'Changed Mind' status
students_who_ever_changed_mind_ids = changed_mind['Student ID'].unique()
# Filter current_grad_info to get all records for these students
all_records_for_changed_students = current_grad_info[current_grad_info['Student ID'].isin(students_who_ever_changed_mind_ids)]
#group the records/program codes by student ID and strip the suffies BA AND convert to list
program_codes_history_per_changed_student = all_records_for_changed_students.groupby('Student ID')['Program Code'].apply(lambda x : x.str.strip('BA.').tolist())
#convert to data frame
pd.DataFrame(program_codes_history_per_changed_student)

,Program Code
Student ID,
1218719,"[UNDC, PSYC, NTH]"
1238033,"[UNDC, IOL, RT]"
1271524,"[UNDC, SO.EN, IPM]"
1310449,"[UNDC, GER.S, PHIL]"
1322044,"[UNDC, RT, R.EN]"
...,...
10941979,"[UNDC, MUS.J, MUS.N]"
10942066,"[UNDC, PH.AS, STR]"
10984039,"[UNDC, PHYS, MA.PH]"


In [72]:
unique_switched_majors_students = changed_mind['Student ID'].nunique()

#what percentage is this in the total number of enrolled students?
#count the number of unique indexes in declared that are in first decisions
total_students = enrollement_and_grad['Student ID'].count()

percent = (unique_switched_majors_students / total_students) * 100
print(f"Out of the {total_students} enrolled students, {unique_switched_majors_students} of them changed their major choices.")
print(f"{int(percent)}% of students switched majors")

Out of the 1628 enrolled students, 247 of them changed their major choices.
15% of students switched majors


In [73]:
def count_switched_in_grad_majors(current_program_code):
  # This 'current_program_code' is a single value from program_codes['Program Code']
  # We want to count how many times this specific program_code appears in first_decisions['Program Code']
  return (changed_mind['Program Code'] == current_program_code).sum()

program_codes['first_choice_programs'] = program_codes['Program Code'].apply(count_switched_in_grad_majors)

In [75]:
changed_mind.nunique()

,0
Student ID,247
Student Status,5
Start Term,5
Program Code,61
Program Status,1
Prog Add Date,200
Prog Status Date,212
Majors at Graduation,78
Currently Declared Majors,13


2. Examining what programs the students swithced to and from

In [81]:
changed_mind['Majors at Graduation'].count()

np.int64(247)

In [93]:
import numpy as np

# Normalize 'Program Code' and 'Majors at Graduation' in changed_mind for consistent comparison
changed_mind_normalized = changed_mind.copy()
changed_mind_normalized['Program Code'] = changed_mind_normalized['Program Code'].str.strip('BA.')
changed_mind_normalized['Majors at Graduation'] = changed_mind_normalized['Majors at Graduation'].str.strip('BA.')

def count_students_in_grad_programs(current_program_code):
  # This 'current_program_code' is a single value from program_codes['Program Code']
  # We want to count how many times this specific program_code appears in first_decisions['Program Code']
  return (changed_mind_normalized['Majors at Graduation'] == current_program_code).sum()


def count_students_in_programs_at_first_choice(current_program):
  return (changed_mind_normalized['Program Code'] == current_program).sum()

program_codes['first_choice_programs'] = program_codes['Program Code'].apply(count_students_in_programs_at_first_choice)
program_codes['grad_majors_count'] = program_codes['Program Code'].apply(count_students_in_grad_programs)
program_codes['diff'] = program_codes['first_choice_programs'] - program_codes['grad_majors_count']

# Calculate percentage_change, handling division by zero using np.where
# If 'first_choice_programs' is 0, the percentage change is 0 (no initial students to change from)
program_codes['percentage_change'] = np.where(
    program_codes['first_choice_programs'] != 0,
    (program_codes['diff'] / program_codes['first_choice_programs']) * 100, 0
)

# Further ensure NaN or inf values resulting from edge cases are set to 0
program_codes['percentage_change'] = program_codes['percentage_change'].replace([np.inf, -np.inf], 0).fillna(0)

program_codes_second_sort = program_codes.sort_values(by = 'first_choice_programs', ascending = False)
program_codes_second_sort

,Program Code,Program Title,normalized_titles,enrollement_interest_Count,first_choice_programs,grad_majors_count,diff,percentage_change
109,PSYC,Psychology,psychology,394,23,12,11,47.826087
41,ECON,Economics,economics,214,17,10,7,41.176471
31,CS,Computer Science,computer science,176,12,8,4,33.333333
25,CHEM,Chemistry,chemistry,181,8,8,0,0.000000
53,GE.EN,Geology-Environmental Studies,geologyenvironmental studies,0,7,2,5,71.428571
...,...,...,...,...,...,...,...,...
137,XHLPC,Health Policy,health policy,0,0,0,0,0.000000
138,XHSSO,"Health, Science, and Society",health science and society,0,0,1,-1,0.000000
139,XNESC,Neuroscience,neuroscience,0,0,0,0,0.000000
140,XPSES,Psychology-Environmental Studies,psychologyenvironmental studies,0,0,2,-2,0.000000


In [95]:
null_codes = np.where((program_codes_second_sort['grad_majors_count'] == 0) & (program_codes_second_sort['first_choice_programs'] == 0) )[0] # retunrs the array insidethe tuple generated by np.where()
print(f"Number of programs with no change in student count {len(null_codes)}")

b_and_a_student_count = program_codes_second_sort.drop(null_codes)


Number of programs with no change in student count 79


In [120]:
display(b_and_a_student_count)

,Program Code,Program Title,normalized_titles,enrollement_interest_Count,first_choice_programs,grad_majors_count,diff,percentage_change
41,ECON,Economics,economics,214,17,10,7,41.176471
31,CS,Computer Science,computer science,176,12,8,4,33.333333
25,CHEM,Chemistry,chemistry,181,8,8,0,0.000000
43,ENGL,English,english,292,5,3,2,40.000000
39,EC.MS,Economics-Mathematics and Statistics,economicsmathematics and statistics,0,3,0,3,100.000000
...,...,...,...,...,...,...,...,...
133,XENFM,Environmental Film,environmental film,0,0,1,-1,0.000000
134,XEPSP,"Environmental Studies, Politics & Sustainable ...",environmental studies politics sustainable pla...,0,0,1,-1,0.000000
138,XHSSO,"Health, Science, and Society",health science and society,0,0,1,-1,0.000000
140,XPSES,Psychology-Environmental Studies,psychologyenvironmental studies,0,0,2,-2,0.000000


In [115]:
program_with_most_switched_from = b_and_a_student_count.sort_values(by='first_choice_programs', ascending=False).iloc[0]
print("Program with the most students initially selecting it (who later changed their mind):")
display(program_with_most_switched_from[['Program Code', 'Program Title', 'first_choice_programs', 'grad_majors_count', 'percentage_change']])

program_with_most_students_grad = b_and_a_student_count.sort_values(by='grad_majors_count', ascending=False).iloc[0]
print("\nProgram with the most students graduating with it (among those who changed their mind):")
display(program_with_most_students_grad[['Program Code', 'Program Title', 'first_choice_programs', 'grad_majors_count', 'percentage_change']])

Program with the most students initially selecting it (who later changed their mind):


,41
Program Code,ECON
Program Title,Economics
first_choice_programs,17
grad_majors_count,10
percentage_change,41.176471



Program with the most students graduating with it (among those who changed their mind):


,41
Program Code,ECON
Program Title,Economics
first_choice_programs,17
grad_majors_count,10
percentage_change,41.176471


This program experienced the most significant decrease in students among those who initially chose it and later changed their major, potentially indicating a trend of students moving away from this field.

This shows the program that had the highest number of students initially selecting it among those who eventually changed their majors.

In [105]:
#STACKED BAR CHART TO REPRESENT THE VARATION IN NO OF STUDENTS IN VARIOUS PROGRAMS BEFORE AND AFTER THEY SWITCHED MAJORS
import plotly.graph_objects as go
fig = go.Figure(data = [
    go.Bar(name='Before students switched majors', x=program_codes_second_sort['first_choice_programs'], y= program_codes_second_sort['Program Code'], orientation='h', marker_color='lightcoral'), # Example color 1
    go.Bar(name='After students switched majors', x= program_codes_second_sort['grad_majors_count'], y= program_codes_second_sort['Program Code'], orientation='h', marker_color='lightseagreen'),
    #hover_data = ('percentage_change')# Example color 2
])

fig.update_layout(
    barmode='stack', # This stacks the bars for each program code
    yaxis={'categoryorder': 'total ascending'}, # Order bars by count in ascending order for better visual flow
    xaxis_title="number of students", # Changed x-axis title to reflect it's now number of students
    yaxis_title="Programs", # Changed y-axis title to reflect it's now programs
    height=max(600, num_programs * 35), # Dynamically adjust height for better visibility of all bars
    margin=dict(l=100, r=20, t=50, b=50)
)
fig.write_html("third_graph.html")
fig.show()

In the plot above displays the variation in student count in different programs before and after students switched majors.
23 students who switched majors had first declared psychology as their major

Determining the rate of graduation for students who never switched majors

In [121]:
# Filter for students who have graduated
graduated_students = enrollement_and_grad[enrollement_and_grad['Student Status'] == 'Graduated'].copy()

# Normalize 'Majors at Graduation' by stripping 'BA.' for consistent comparison
graduated_students['normalized_grad_major'] = graduated_students['Majors at Graduation'].str.strip('BA.').fillna('')

# Define a function to check if the graduated major is in the initial fuzzy interests
def stuck_with_interest(row):
    grad_major = row['normalized_grad_major']
    fuzzy_interests = row['fuzzy_interest_codes']
    return grad_major in fuzzy_interests and grad_major != ''

# Apply the function to count students who stuck with their interest
students_stuck_with_first_year_interest = graduated_students['normalized_grad_major'].apply(lambda x: x if x in program_map.keys() else None).dropna().index.size

# Count students where the final graduated major was among their initial fuzzy interests
count_stuck = graduated_students.apply(stuck_with_interest, axis=1).sum()

# Calculate the total number of graduated students for percentage
total_graduated_students = len(graduated_students)

# Calculate the percentage
percentage_stuck = (count_stuck / total_graduated_students) * 100 if total_graduated_students > 0 else 0

print(f"Out of {total_graduated_students} graduated students, {count_stuck} stuck with one of their first-year academic interests as their graduation major.")
print(f"This is {percentage_stuck:.2f}% of all graduated students.")

Out of 1562 graduated students, 367 stuck with one of their first-year academic interests as their graduation major.
This is 23.50% of all graduated students.


In [52]:
program_codes_per_student = changed_mind.groupby('Student ID')['Program Code'].apply(list)
print("Program codes for each student who changed majors:")
program_codes_per_student

Program codes for each student who changed majors:


,Program Code
Student ID,
1218719,[BA.PSYC]
1238033,[BA.BIOL]
1271524,[BA.SO.EN]
1310449,[BA.GER.S]
1322044,[BA.ART]
...,...
10941979,[BA.MUS.J]
10942066,[BA.PH.AS]
10984039,[BA.PHYS]


# Conclusion


**So how many students within 2017 and 2021 switched majors?**

After filtering out the students with major changes, I discovered that 247 (15%) out of the recorded 1628  students had switched majors at least once during their three years in college.

**Which mjors did they switch to and from?**

The third visualization shows the trend in student counts across various programs. Notably, Economics, Computer Science, and Psychology experienced the largest percentage decline(between 40% and 42%) n the number of students who switched majors.

**Of all the ones who graduated, how many actually stuck with the interests they made from first year?**

Out of 1562 graduating students, 367 chose one of their first-year academic interests as their major.
This is 23.50% of all graduated students.